Add support for uploading to a S3 AWS bucket.

Bluebugs opened the pull request.

Commit:
Add support for uploading to a S3 AWS bucket.

Commit:
Handle empty template in selfupdatectl upload.

Commit:
Handle .Executable in http_source template to match selfupdatectl upload template.

andydotxyz commented on cmd/selfupdatectl/internal/cloud/aws.go:
```diff
@@ -0,0 +1,131 @@
+
+	atomic.AddInt64(&pa.downloaded, int64(n))
+
+	fmt.Printf("\r%v: %v%% %c", pa.file, 100*pa.downloaded/(pa.contentLength*2), pa.tick())
```
content Length *2? why is this?
As this isn't using terminal commands won't it just draw the progress ticker on lots of lines?
I may be misunderstanding as I haven't run the command, but I don't want to spam the console.

Bluebugs replied:
The formula is the one provided by aws go sdk. They apparently read the file twice while sending it, so it needs to be accounted for.

As for the drawing, it is using a \r (return cariage) at the beginning of the line, which means it is not going to the next line, but erasing the last line not finished by \n that was printed. Basically erasing the same line.

andydotxyz replied:
This works on Windows cmd as well?

Bluebugs replied:
Funnily enough this is so old that this behavior predate existence of Windows. It would work on the oldest terminal as it only rely on carriage return and return line.

andydotxyz replied:
I just wanted to check as my knowledge of windows and line returns etc is just pain!

Bluebugs replied:
I understand, I had to check that it worked in practice and it was not just a theory. A bit to simple of a trick to be believed :-)

andydotxyz commented on cmd/selfupdatectl/upload.go:
```diff
@@ -0,0 +1,180 @@
+				Destination: &baseS3Path,
+			},
+			&cli.StringFlag{
+				Name:        "binary-template",
```
This name confuses me a little, as everywhere else it is called executable, not binary.
Maybe "template" is enough.
One thing I am not sure about is if the subcommand "upload" is S3 specific, if so if we may add other upload targets maybe it should be "s3upload".
If this is indeed aws s3 specific then all of the parameters having "aws" prefix seems like duplication (and maybe the s3 mention in base path too).

Bluebugs replied:
It is s3 specific at the moment, but I am hoping that we can have parameters to set different upload mode and have one name. I can easily rename to s3upload for the moment and when we add new mode we can merge them then.

andydotxyz commented on cmd/selfupdatectl/upload.go:
```diff
@@ -0,0 +1,180 @@
+				return fmt.Errorf("at least one executable to upload")
+			}
+
+			log.Println("Connecting to AWS")
```
This function is nearly 150 lines long. Once the arg parsing is done can it be factored into more testable / readable functions?

andydotxyz commented on cmd/selfupdatectl/upload.go:
```diff
@@ -0,0 +1,180 @@
+				}
+			}
+
+			for _, exe := range ctx.Args().Slice() {
```
As above it seems that this could be factored out to a function that uploads one thing, which is called in this loop?

andydotxyz commented on http_source.go:
```diff
@@ -120,6 +122,16 @@ func replaceUrlTemplate(base string) string {
 	}
 
+	exe, err := getExecutableRealPath()
+	if err != nil {
```
Doesn't this belong in `getExecutableRealPath`?
Then if the error is returned we cannot ignore it...

Bluebugs replied:
Actually I made a mistake here, the idea is I just need the executable filename, not its full path. So I can fallback to os.Args[0] something that getExecutableRealPath can not.

andydotxyz reviewed (changes requested):
This PR is basically setting up some UI and I think we should discuss the scope of "upload" vs "s3upload" and other aspects of how it should work.

For example this assumes that GOOS is the os of the executable passed in, but as this is an uploader not a builder it will probably be called many times for different files. The template really helps with this, but it is lacking the ability to specify os in the parameters, falling back to the GOOS env, which may not be obvious.

Commit:
Use Base to only return the executable filename and not its full path.

Bluebugs commented:
> For example this assumes that GOOS is the os of the executable passed in, but as this is an uploader not a builder it will probably be called many times for different files. The template really helps with this, but it is lacking the ability to specify os in the parameters, falling back to the GOOS env, which may not be obvious.

My usage guess would be that the uploader will be called multiple time with each file being for a different target OS, but for the same application. So uploading one executable can be helped with an automatic template as that would be the likely use case. With multiple upload at the same time, my expectation would be that the caller already has named the file properly during his build process to avoid mistake and so no rename is expected. Which means likely disabling the template in that case.

The other upload form I had considered, would be more classic: `selfupdatectl upload file file file targetpath`. This would remove completely the need for the template and let the user do all the renaming.

andydotxyz commented:
> My usage guess would be that the uploader will be called multiple time with each file being for a different target OS, but for the same application. So uploading one executable can be helped with an automatic template as that would be the likely use case.

Yes I agree with this. But how do we pass in the OS/arch that was used to create this file? It won't matter with just the filename specified but if we are to support templating then this will pick up the runtime not compile time settings.

Commit:
For better clarity of command line, split multi upload from single upload.

This also add support for overriding the runtime GOOS and GOARCH case
when doing single file upload.

Bluebugs commented:
> Yes I agree with this. But how do we pass in the OS/arch that was used to create this file? It won't matter with just the filename specified but if we are to support templating then this will pick up the runtime not compile time settings.

I think the easiest is to split the case where you would want to override it, which is single file upload and when you wouldn't. I have split to two command, `s3upload` which use a template and `s3uploads` that doesn't. It does lead to a more logical command line experience I think.

Bluebugs commented:
Oh, I realize I should be updating the README.md in this PR now. Will do.

Commit:
Merge branch 'main' into features/aws-s3-integration

Commit:
Update documentation.

Commit:
Simplify command line.

Commit:
Update documentation

Commit:
Ooops remove unused code.

andydotxyz commented on cmd/selfupdatectl/upload.go:
```diff
@@ -0,0 +1,133 @@
+
+	return &cli.Command{
+		Name:        "s3upload",
+		Usage:       "Upload multiple executable which has been properly signed to a S3 path",
```
I think this doc is not right - it refers to multiple executables.
Also it could read as "properly signed" being the input, not the result - perhaps "Upload an executable file to S3, it will be signed and the signature uploaded too"?

Commit:
Update command comments accordingly.

Commit:
Correct gender.

andydotxyz reviewed (approved):
S3-tastic, thanks
